// react-maplibre/src/stories/MapLines.stories.tsx
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Map from '../component/Map';
import { MaplibreProps } from "../component/Map/Map.types"
import {  
  MapOption,
  Line,
  Point
} from "../types"
import { useState } from '@storybook/addons';
import { nanoid } from 'nanoid'
import * as MapColor from "./../constant/color"
import { decode }  from "./../util/polyline"


export default {
  title: 'Example/MapLines',
  component: Map,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof Map>;


const defaultMapOptions:MapOption = {
  zoom: 12,
  center: [120.9991044, 24.8430631],
  style: "https://service.polstarapis.com/mapstyle/default/style.json",
  touchZoomRotateDisable: true,
  cursorIcon: "default"
}

const MapDrawLinesTemplate: ComponentStory<typeof Map> = 
  (args:MaplibreProps) => 
{
  const polyline = "mbsvCoroaVc@bAfAf@bBx@~@{BZs@l@sAl@uArAyDr@oBjBmFHSdA{CVu@dCmHFShC_H`BaEnB_FrDcHbA}ANa@Lc@Tu@Ps@Pw@RsBJ_E?oB?aCHuGByE?aCEmAWoA_AkEJi@XYdA?`HrCpCbAHUx@sC"

  const [lines,setLines] = useState<Line[]>([
    {
      priority: 1,
      color: MapColor.LINE_BLUE,
      outlineColor: MapColor.OUTLINE_BLUE,
      opacity: 0.5,
      coordinates:decode(polyline).map((coordinate)=>coordinate.reverse())
    }
  ])
  const [points,setPoints] = useState<Point[]>([
    {
      id: nanoid(),
      num:0,
      lngLat: [120.9991044,24.8430631],
      img:"./blue_border_poi.svg",
      style:{
        strokeShadow:`
          -1px 1px 0 #005598,
          1px 1px 0 #005598,
          1px -1px 0 #005598,
          -1px -1px 0 #005598;
        `
      }
    },
    {
      id: nanoid(),
      num:1,
      lngLat: [121.0197991,24.8326827],
      img:"./blue_border_poi.svg",
      style:{
        strokeShadow:`
          -1px 1px 0 #005598,
          1px 1px 0 #005598,
          1px -1px 0 #005598,
          -1px -1px 0 #005598;
        `
      }
    }
  ])


  return <Map 
    points={points}
    lines={lines} 
    option={defaultMapOptions}
    onMapLoad={()=>{
    }}
  />;
} 


export const MapDrawLines = MapDrawLinesTemplate.bind({});
