// react-maplibre/src/stories/MapPoints.stories.tsx
import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import Map from '../component/Map';
import { MaplibreProps } from "../component/Map/Map.types"
import {  
  MapOption,
  Point,
  Popup
} from "../types"
import { useState } from '@storybook/addons';
import maplibregl,{ 
  Map as MaplibreMap,
  MapLibreEvent,
  MapMouseEvent, 
} from "maplibre-gl"
import { nanoid } from 'nanoid'


export default {
  title: 'Example/MapPoints',
  component: Map,
  argTypes: {
    backgroundColor: { control: 'color' },
  },
} as ComponentMeta<typeof Map>;


const defaultMapOptions:MapOption = {
  zoom: 15,
  center: [121.5333742, 25.054579],
  style: "https://service.polstarapis.com/mapstyle/default/style.json",
  touchZoomRotateDisable: true,
  cursorIcon: "default"
}

const MapDrawPointsTemplate: ComponentStory<typeof Map> = 
  (args:MaplibreProps) => 
{
  const [points,setPoints] = useState<Point[]>([])

  const handleMapClick = (ev:MapMouseEvent,map:MaplibreMap)=>{
    const point:Point = {
      id: nanoid(),
      num:points.length + 1,
      lngLat: [ev.lngLat.lng,ev.lngLat.lat],
      img:"./blue_border_poi.svg",
      style:{
        strokeShadow:`
          -1px 1px 0 #005598,
          1px 1px 0 #005598,
          1px -1px 0 #005598,
          -1px -1px 0 #005598;
        `
      }
    }
    setPoints([...points,point])
  }

  return <Map 
    points={points} 
    option={defaultMapOptions}
    onMapClick={handleMapClick}
    onMapLoad={()=>{
    }}
  />;
} 

const MapDrawPointsWithPopupTemplate: ComponentStory<typeof Map> = 
  (args:MaplibreProps) => 
{
  const [points,setPoints] = useState<Point[]>([])

  const handleMapClick = (ev:MapMouseEvent,map:MaplibreMap)=>{

    const popupContent = document.createElement("div")
    popupContent.innerHTML = "hell world"

    const msgPopup: Popup = {
        domContent: popupContent,
        lngLat: [ev.lngLat.lng,ev.lngLat.lat],
        option: {
            closeButton: false,
            closeOnClick: false,
            anchor: "bottom"
        }
    }

    const point:Point = {
      id: nanoid(),
      num:points.length + 1,
      lngLat: [ev.lngLat.lng,ev.lngLat.lat],
      img:"./blue_border_poi.svg",
      popup:msgPopup,
      style:{
        strokeShadow:`
          -1px 1px 0 #005598,
          1px 1px 0 #005598,
          1px -1px 0 #005598,
          -1px -1px 0 #005598;
        `
      }
    }

    setPoints([...points,point])
  }

  return <Map 
    points={points} 
    option={defaultMapOptions}
    onMapClick={handleMapClick}
    onMapLoad={()=>{
    }}
  />;
} 


export const MapDrawPoints = MapDrawPointsTemplate.bind({});
export const MapDrawPointsWithPopup = MapDrawPointsWithPopupTemplate.bind({});
