// react-maplibre/src/util/geojson.tsx
import { SourceSpecification } from "maplibre-gl"
import * as Type from "../types"

export const createCirclesGeoJson = (circles:Type.Circle[]): SourceSpecification => {
    const fetures = circles.map((circle) => {
        const distanceX = circle.radius / (111.320 * Math.cos(circle.lngLat[1] * Math.PI / 180));
        const distanceY = circle.radius  / 110.574;

        const points = circle.point ? circle.point : 128
        const coordinates:any[][] = []

        for(let i = 0;i < points ;i++){
            const theta = (i / points) * (2 * Math.PI);
            const x = distanceX * Math.cos(theta)
            const y = distanceY * Math.sin(theta)
            coordinates.push([
                circle.lngLat[0] + x,
                circle.lngLat[1] + y
            ])
        }

        coordinates.push(coordinates[0])

        return {
            "type": "Feature",
            "geometry": {
                "type": "Polygon",
                "coordinates": [coordinates]
            }
        }
    })

    return {
        "type": "geojson",
        "data": {
            "type": "FeatureCollection",
            "features": fetures
        },
    };
};


export const createRectangsleGeoJson = (rectanges:Type.Rectangle[]): SourceSpecification => {
    
    const features = rectanges.map((rectange)=>{
        const coordinate1: [number, number] = [rectange.coordinates[0][0], rectange.coordinates[1][1]]
        const coordinate2: [number, number] = [rectange.coordinates[1][0], rectange.coordinates[0][1]]

        return {
            "type": "Feature",
            "geometry": {
                "type": "Polygon",
                "coordinates": [[rectange.coordinates[0], coordinate1, rectange.coordinates[1], coordinate2, rectange.coordinates[0]]]
            }
        }
    })
    
    return {
        "type": "geojson",
        "data": {
            "type": "FeatureCollection",
            "features":features
        },
    };
}

export const createLinesGeoJson = (lines:Type.Line[]):SourceSpecification=>{

    const sortLines = lines.sort((a,b) => ( a?.priority || 0 ) - ( b?.priority || 0))

    const fetures = sortLines.map((line) => {
        return {
            "type": "Feature",
            "properties": {
                'color':line.color,
                'opacity':line.opacity,
                'outlineColor':line.outlineColor
            },
            "geometry": {
                "type": "LineString",
                "coordinates": line.coordinates
            }
        }
    })

    return {
        type: 'geojson',
        data: {
            "type": "FeatureCollection",
            "features": fetures
        }
    }
}

export const createLineGeoJson = (line:Type.Line):SourceSpecification=>{
    return {
        type: 'geojson',
        data: {
            "type": "FeatureCollection",
            "features": {
                "type": "Feature",
                "properties": {},
                "geometry": {
                    "type": "LineString",
                    "coordinates": line.coordinates
                }
            }
        }
    }
}
