// react-maplibre/src/util/render.tsx
import * as Type from "../types"
import maplibregl, { LngLatLike, Map as MaplibreMap, Marker, Popup, SourceSpecification, PopupOptions, MapMouseEvent } from "maplibre-gl"
import { createLinesGeoJson, createCirclesGeoJson, createRectangsleGeoJson, createLineGeoJson } from "./geojson"
import * as Fire from "./../constant/fire"
import * as Color from "./../constant/color"


const generatePopup = (popup:Type.Popup): Popup => {
    const newPopup = new maplibregl.Popup(popup.option).setHTML(popup.html ? popup.html : "")
    if(popup.domContent){
        newPopup.setDOMContent(popup.domContent)
    }
    return newPopup
}

const generateMarker = (point:Type.Point): Marker => {
    const pointHtml = point.html ? point.html : `<div 
            class="map-point"
            style="
                position:relative;
                display: flex;
                align-items: center;
                justify-content: center;
                user-select: none;
                width:100%;
                height:100%;
            "
        >
            <img
                class="map-point-img"
                style="
                    filter: drop-shadow(1px 1px 4px rgba(0, 0, 0, 0.15));
                    width:100%;
                    height:100%;
                    object-fit:contain;
                "
                src="${point.img ? point.img : "/icon/fill_blue_poi.svg"}"
            />
            <span 
                class="map-point-num"
                    style="
                    position: absolute;
                    z-index: 3;
                    color: white;
                    font-weight:bold;
                    font-size: 14px;
                    text-shadow:${point.style?.strokeShadow ? point.style?.strokeShadow : "none"};
                    margin-bottom:3px;
                "
            >
                ${point.num != undefined ? point.num : ""}
            </span>
        </div>
    `
    const marker = new maplibregl.Marker({
        element: new DOMParser().parseFromString(pointHtml,"text/html").body
    })

    if(point.popup){
        marker.setPopup(generatePopup(point.popup))
    }

    return marker;
}

export const popupsRendering = (popups: Type.Popup[], map: MaplibreMap) => {
    map.fire(Fire.REMOVE_POPUPS)
    for (let i = 0; i < popups.length; i++) {
        const p = generatePopup(popups[i])
        p.setLngLat(popups[i].lngLat).addTo(map)

        map.once(Fire.REMOVE_POPUPS, () => {
            p.remove()
        })

        const onClickCallback = popups[i].onClick
        if(onClickCallback){
            p.getElement().addEventListener("click",(e)=>{
                onClickCallback(e)
            })
        }

        const onMouseOverCallback = popups[i].onMouseOver
        if(onMouseOverCallback){
            p.getElement().addEventListener("mouseover",(e)=>{
                onMouseOverCallback(e)
            })
        }

        const onMouseout = popups[i].onMouseout
        if(onMouseout){
            p.getElement().addEventListener("mouseout",(e)=>{
                onMouseout(e)
            })
        }

        if (!popups[i].zoom) continue

        if (map.getZoom() < 10.98) {
            p.setHTML("")
        }

        map.on("zoomstart", (event: MapMouseEvent) => {
            if (!popups[i].zoom) return
            p.setHTML("")
        })

        map.on("zoomend", (event: MapMouseEvent) => {
            if (map.getZoom() < 10.98) {
                p.setHTML("")
            } else {
                const content = popups[i].html ? popups[i].html : ""
                if (content) {
                    p.setHTML(content)
                }
            }
        })
    }
    return map
}

export const pointsRendering = (points: Type.Point[], map: MaplibreMap) => {
    map.fire(Fire.REMOVE_POINTS)

    const sortPoints = points.sort((a,b) => ( a?.priority || 0 ) - ( b?.priority || 0))

    for (let i = 0; i < sortPoints.length; i++) {
        if (sortPoints[i]?.hidden) continue

        sortPoints[i].style

        const pointMarker = generateMarker(sortPoints[i])

        const onClickCallback = sortPoints[i].onClick
        if(onClickCallback){
            pointMarker.getElement().addEventListener("mousedown",(e:any)=>{
                onClickCallback(e,pointMarker)
            })
        }

        const onMouseOverCallback = sortPoints[i].onMouseOver
        if(onMouseOverCallback){
            pointMarker.getElement().addEventListener("mouseover",(e)=>{
                onMouseOverCallback(e,pointMarker)
            })
        }

        const onMouseout = sortPoints[i].onMouseout
        if(onMouseout){
            pointMarker.getElement().addEventListener("mouseout",(e)=>{
                onMouseout(e,pointMarker)
            })
        }

        pointMarker.setLngLat(sortPoints[i].lngLat).addTo(map)

        map.once(Fire.REMOVE_POINTS, () => {
            pointMarker.remove()
        })
    }

    return map
}

export const linesRendering = (
    lines: Type.Line[],
    map: MaplibreMap,
) => {
    map.fire(Fire.REMOVE_LINES)

    map.once(Fire.REMOVE_LINES, () => {
        if (map.getLayer(`linesLayer`)) {
            map.removeLayer(`linesLayer`);
        }
        if (map.getLayer(`linesBorderLayer`)) {
            map.removeLayer(`linesBorderLayer`);
        }
        if (map.getSource(`linesSource`)) {
            map.removeSource(`linesSource`)
        }
    })

    const source = createLinesGeoJson(lines)

    map.addSource("linesSource", source);

    map.addLayer(
        {
            'id': `linesBorderLayer`,
            'type': 'line',
            'source': "linesSource",
            'layout': {
                "line-join": "round",
                "line-cap": "round",
            },
            'paint': {
                'line-color': [
                    'match',
                    ['get', 'outlineColor'],
                    Color.OUTLINE_BLUE,
                    Color.OUTLINE_BLUE,
                    Color.OUTLINE_GRAY,
                    Color.OUTLINE_GRAY,
                    Color.OUTLINE_BLUE,
                ],
                'line-width': 1,
                'line-gap-width': 4,
            }
        },
    );

    map.addLayer(
        {
            'id': "linesLayer",
            'type': 'line',
            'source': "linesSource",
            'layout': {
                "line-join": "round",
                "line-cap": "round",
            },
            'paint': {
                'line-color': [
                    'match',
                    ['get', 'color'],
                    Color.LINE_BLUE,
                    Color.LINE_BLUE,
                    Color.LINE_GRAY,
                    Color.LINE_GRAY,
                    Color.LINE_BLUE,
                ],
                'line-width': 6,
                'line-opacity': ['get', 'opacity'],
            }
        }
    );

    return map
}

export const circlesRendering = (circles: Type.Circle[], map: MaplibreMap) => {
    map.fire(Fire.REMOVE_CIRCLES)

    const source = createCirclesGeoJson(circles)

    map.once(Fire.REMOVE_CIRCLES, () => {
        if (map.getLayer("circleLayer")) {
            map.removeLayer("circleLayer");
        }

        if (map.getLayer(`circleOutline`)) {
            map.removeLayer(`circleOutline`);
        }

        if (map.getSource('circleOutlineSource')) {
            map.removeSource('circleOutlineSource')
        }

        if (map.getSource('circleSource')) {
            map.removeSource('circleSource')
        }
    })

    map.addSource('circleOutlineSource', source);
    map.addSource('circleSource', source);

    map.addLayer({
        "id": "circleLayer",
        "type": "fill",
        "source": "circleSource",
        "layout": {},
        "paint": {
            "fill-color": "#1976D2",
            "fill-opacity": 0.1,
        }
    });

    map.addLayer(
        {
            'id': 'circleOutline',
            'type': 'line',
            'source': 'circleOutlineSource',
            'layout': {
                "line-join": "round",
                "line-cap": "round",
            },
            'paint': {
                'line-color': '#1976D2',
                'line-width': 1,
            },
        }
    );

    return map
}


export const rectanglesRendering = (rectangles: Type.Rectangle[], map: MaplibreMap) => {
    map.fire(Fire.REMOVE_RECTANGELS)

    const source = createRectangsleGeoJson(rectangles)

    map.once(Fire.REMOVE_RECTANGELS, () => {
        if (map.getLayer("rectanglesLayer")) {
            map.removeLayer("rectanglesLayer");
        }

        if (map.getLayer(`rectanglesOutline`)) {
            map.removeLayer(`rectanglesOutline`);
        }

        if (map.getSource('rectanglesOutlineSource')) {
            map.removeSource('rectanglesOutlineSource')
        }

        if (map.getSource('rectanglesSource')) {
            map.removeSource('rectanglesSource')
        }
    })

    map.addSource('rectanglesOutlineSource', source);
    map.addSource('rectanglesSource', source);

    map.addLayer({
        "id": "rectanglesLayer",
        "type": "fill",
        "source": "rectanglesSource",
        "layout": {},
        "paint": {
            "fill-color": "#1976D2",
            "fill-opacity": 0.1,
        }
    });

    map.addLayer(
        {
            'id': 'rectanglesOutline',
            'type': 'line',
            'source': 'rectanglesOutlineSource',
            'layout': {
                "line-join": "round",
                "line-cap": "round",
            },
            'paint': {
                'line-color': '#1976D2',
                'line-width': 1,
            },
        }
    );

    return map
}

const polygonRendering = (
    source: SourceSpecification,
    layerName: string,
    sourceName: string,
    map: MaplibreMap
) => {

    map.on("removePolygon", () => {
        if (map.getLayer(layerName)) {
            map.removeLayer(layerName);
        }

        if (map.getLayer(`${sourceName}-border`)) {
            map.removeLayer(`${sourceName}-border`);
        }

        if (map.getSource('polygonOutlineSource')) {
            map.removeSource('polygonOutlineSource')
        }

        if (map.getSource('polygonSource')) {
            map.removeSource('polygonSource')
        }
    })

    map.fire("removePolygon")
    map.addSource('polygonOutlineSource', source);
    map.addSource('polygonSource', source);

    map.addLayer({
        "id": "polygonLayer",
        "type": "fill",
        "source": "polygonSource",
        "layout": {},
        "paint": {
            "fill-color": "#1976D2",
            "fill-opacity": 0.1,
        }
    });

    map.addLayer(
        {
            'id': 'polygonOutline',
            'type': 'line',
            'source': 'polygonOutlineSource',
            'layout': {
                "line-join": "round",
                "line-cap": "round",
            },
            'paint': {
                'line-color': '#1976D2',
                'line-width': 1,
            },
        }
    );
}
